File: SDK/typescript/sarembok_sdk.ts
/**
 * Sarembok VE Production Edition — Official TypeScript SDK
 * Exposes all 12 Platform RPC Facets for External Web/Node Integrations.
 */

export class SarembokClient {
    private url: string;
    private ws: WebSocket | null = null;

    constructor(host: string = "127.0.0.1", port: number = 9000) {
        this.url = `ws://${host}:${port}`;
    }

    public async connect(): Promise<boolean> {
        return new Promise((resolve, reject) => {
            this.ws = new WebSocket(this.url);
            this.ws.onopen = () => resolve(true);
            this.ws.onerror = (err) => reject(err);
        });
    }

    private async call(method: string, params: Record<string, any> = {}): Promise<any> {
        if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
            await this.connect();
        }
        return new Promise((resolve) => {
            const id = `sdk-${method.toLowerCase()}`;
            const handler = (event: MessageEvent) => {
                const data = JSON.parse(event.data);
                if (data.id === id) {
                    this.ws?.removeEventListener("message", handler);
                    resolve(data);
                }
            };
            this.ws?.addEventListener("message", handler);
            this.ws?.send(JSON.stringify({ jsonrpc: "2.0", id, method, params }));
        });
    }

    public async createAgent(agentId: string, displayName: string) {
        return this.call("CreateAgent", { agentId, displayName });
    }

    public async queryAgentState(agentId: string) {
        return this.call("QueryAgentState", { agentId });
    }

    public async injectPerception(agentId: string, perceptionJson: string) {
        return this.call("InjectPerception", { agentId, perception: perceptionJson });
    }

    public async evaluateDecision(agentId: string, actionId: string, riskScore: number, confidence: number) {
        return this.call("EvaluateDecision", { agentId, actionId, riskScore, confidence });
    }

    public async getCognitiveScorecard(agentId: string) {
        return this.call("GetCognitiveScorecard", { agentId });
    }

    public async queryWorldModel(filter: string = "all") {
        return this.call("QueryWorldModel", { filter });
    }

    public async createDelegation(sourceAgentId: string, targetAgentId: string, goalId: string) {
        return this.call("CreateDelegation", { sourceAgentId, targetAgentId, goalId });
    }

    public async getAuditTrail(agentId: string) {
        return this.call("GetAuditTrail", { agentId });
    }

    public async sendMessage(agentId: string, content: string) {
        return this.call("SendMessage", { agentId, content });
    }

    public async getEvents(agentId: string) {
        return this.call("GetEvents", { agentId });
    }

    public async getMetrics(agentId: string) {
        return this.call("GetMetrics", { agentId });
    }

    public async restoreState(agentId: string, walEntries: number = 0) {
        return this.call("RestoreState", { agentId, walEntries });
    }

    public close() {
        this.ws?.close();
    }
}

File: SDK/python/sarembok_sdk.py
"""
Sarembok VE Production Edition — Official Python SDK
Exposes all 12 Platform RPC Facets for External Integrations.
"""

import json
import websocket

class SarembokClient:
    def __init__(self, host="127.0.0.1", port=9000):
        self.url = f"ws://{host}:{port}"
        self.ws = None

    def connect(self):
        self.ws = websocket.create_connection(self.url)
        return True

    def _call(self, method, params=None):
        if not self.ws:
            self.connect()
        payload = {
            "jsonrpc": "2.0",
            "id": f"sdk-{method.lower()}",
            "method": method,
            "params": params or {}
        }
        self.ws.send(json.dumps(payload))
        resp = self.ws.recv()
        return json.loads(resp)

    # 1. Agent Management
    def create_agent(self, agent_id, display_name):
        return self._call("CreateAgent", {"agentId": agent_id, "displayName": display_name})

    # 2. Agent State
    def query_agent_state(self, agent_id):
        return self._call("QueryAgentState", {"agentId": agent_id})

    # 3. Perception
    def inject_perception(self, agent_id, perception_json):
        return self._call("InjectPerception", {"agentId": agent_id, "perception": perception_json})

    # 4. Decisions & Governance
    def evaluate_decision(self, agent_id, action_id, risk_score, confidence):
        return self._call("EvaluateDecision", {
            "agentId": agent_id, "actionId": action_id,
            "riskScore": risk_score, "confidence": confidence
        })

    # 5. Scorecard
    def get_cognitive_scorecard(self, agent_id):
        return self._call("GetCognitiveScorecard", {"agentId": agent_id})

    # 6. World Model
    def query_world_model(self, filter_str="all"):
        return self._call("QueryWorldModel", {"filter": filter_str})

    # 7. Delegation
    def create_delegation(self, source_agent_id, target_agent_id, goal_id):
        return self._call("CreateDelegation", {
            "sourceAgentId": source_agent_id,
            "targetAgentId": target_agent_id,
            "goalId": goal_id
        })

    # 8. Audit Trail
    def get_audit_trail(self, agent_id):
        return self._call("GetAuditTrail", {"agentId": agent_id})

    def close(self):
        if self.ws:
            self.ws.close()
